File: motor-controller/common/git.c
/**
 * @file git.c
 * @author Nicholas Witten
 * @brief Struct instantiation to hold embeddable git information
 * @version 0.1
 * @date 2022-05-22
 * 
 * @copyright Copyright (c) 2022
 * 
 */

#include "git.h"


static volatile GitStatus_t git_status = {0xAABBCCDD, 0, 0};  // Git hash and dirty will be embedded upon build


uint32_t get_git_hash(void) {
    return git_status.git_hash;
}

uint32_t get_git_dirty(void) {
    return git_status.git_dirty;
}

// Keeps compiler from optimizing this field out
uint32_t get_git_struct_id(void) {
    return git_status.git_struct_id;
}

File: control-board/src/git.rs
struct GitStatus {
    git_struct_id: u32,
    git_hash: u32,
    git_dirty: u32,
}
const GIT_STATUS: GitStatus = GitStatus {git_struct_id: 0xAABBCCDD, git_hash: 0, git_dirty: 0};

pub fn get_git_id() -> u32 {
    // Enforce a read from memory
    unsafe {
        let git_status_ptr: *const GitStatus = &GIT_STATUS;
        return (*git_status_ptr).git_struct_id;
    }
}

pub fn get_git_hash() -> u32 {
    // Enforce a read from memory
    unsafe {
        let git_status_ptr: *const GitStatus = &GIT_STATUS;
        return (*git_status_ptr).git_hash;
    }
}

pub fn get_git_dirty() -> u32 {
    // Enforce a read from memory
    unsafe {
        let git_status_ptr: *const GitStatus = &GIT_STATUS;
        return (*git_status_ptr).git_dirty;
    }
}